web-common/src/features/models/createDashboardFromModel.ts

```diff
@@ -0,0 +1,213 @@
+import { derived, get } from "svelte/store";
+
+export interface CreateDashboardFromModelRequest {
+  instanceId: string;
+  modelName: string;
+  schema: V1StructType;
+  newDashboardName: string;
+}
+
+export const useCreateDashboardFromModel = <
+  TError = RpcStatus,
+  TContext = unknown
+>(options?: {
+  mutation?: CreateMutationOptions<
+    Awaited<Promise<void>>,
+    TError,
+    { data: CreateDashboardFromModelRequest },
+    TContext
+  >;
+}) => {
+  const { mutation: mutationOptions } = options ?? {};
+  const queryClient = mutationOptions?.queryClient ?? useQueryClient();
+
+  const mutationFn: MutationFunction<
+    Awaited<Promise<void>>,
+    { data: CreateDashboardFromModelRequest }
+  > = async (props) => {
+    const { data } = props ?? {};
+    if (!data.schema)
+      throw new Error("Failed to create dashboard: Model is not ready");
+
+    // create dashboard from model
+    const dashboardYAML = generateDashboardYAMLForModel(
+      data.modelName,
+      data.schema,
+      data.newDashboardName
+    );
+
+    await runtimeServicePutFile(
+      data.instanceId,
+      getFileAPIPathFromNameAndType(
+        data.newDashboardName,
+        EntityType.MetricsDefinition
+      ),
+      {
+        blob: dashboardYAML,
+        create: true,
+        createOnly: true,
+      }
+    );
+    await waitForResource(
+      queryClient,
+      data.instanceId,
+      getFilePathFromNameAndType(
+        data.newDashboardName,
+        EntityType.MetricsDefinition
+      )
+    );
+  };
+
+  return createMutation<
+    Awaited<Promise<void>>,
+    TError,
+    { data: CreateDashboardFromModelRequest },
+    TContext
+  >(mutationFn, mutationOptions);
+};
+
+function modelHasDataStore(
+  queryClient: QueryClient,
+  instanceId: string,
+  modelName: string
+) {
+  return derived(
+    [
+      useResource(
+        instanceId,
+        modelName,
+        ResourceKind.Model,
+        undefined,
+        queryClient
+      ),
+      createSchemaForTable(
+        instanceId,
+        modelName,
+        ResourceKind.Model,
+        queryClient
+      ),
+    ],
+    ([model, schema]) => {
+      return (
+        !model.isFetching &&
+        !!model.data &&
+        model.data.meta.reconcileStatus ===
+          V1ReconcileStatus.RECONCILE_STATUS_IDLE &&
+        !schema.isFetching &&
+        !!schema.data
+      );
+    }
+  );
+}
+
+function waitForModelHasData(
+  queryClient: QueryClient,
+  instanceId: string,
+  modelName: string
+) {
+  const store = modelHasDataStore(queryClient, instanceId, modelName);
+  if (get(store)) return;
+  return new Promise<void>((resolve) => {
+    const unsub = store.subscribe((hasData) => {
+      if (!hasData) return;
+      resolve();
+      unsub();
+    });
+  });
+}
+
+/**
+ * Wrapper function that takes care of UI side effects on top of creating a dashboard from model.
+ * TODO: where would this go?
+ */
+export function useCreateDashboardFromModelUIAction(
```

Comment:
Nice, this is great decoupling the UI logic from the action logic. I think this is a reasonable place to put the code.